tsl/src/compression/compression.c

```diff
@@ -1986,9 +2042,78 @@ create_segment_filter_scankey(RowDecompressor *decompressor, char *segment_filte
 	return num_scankeys;
 }
 
+/*
+ * Helper method to build scankeys to do lookup into compressed chunk index
+ */
+static ScanKeyData *
+fetch_index_scankeys(const Relation idxrel, Oid hypertable_relid, TupleTableSlot *uncompressedslot,
+					 int *num_keys, const int key_flags)
+{
+	ScanKeyData *scankey = NULL;
+	int indnkeyatts;
+	int total_keys = 0;
+	int attoff;
+	bool isnull;
+	Datum indclassDatum;
+	oidvector *opclass;
+
+	indclassDatum =
+		SysCacheGetAttr(INDEXRELID, idxrel->rd_indextuple, Anum_pg_index_indclass, &isnull);
+	Assert(!isnull);
+	opclass = (oidvector *) DatumGetPointer(indclassDatum);
+
+	indnkeyatts = IndexRelationGetNumberOfKeyAttributes(idxrel);
+	scankey = palloc0(sizeof(ScanKeyData) * indnkeyatts);
+
+	/* Build scankey for every attribute in the index. */
+	for (attoff = 0; attoff < indnkeyatts; attoff++)
+	{
+		Oid operator;
+		Oid opfamily;
+		RegProcedure regop;
+		int pkattno = attoff + 1;
+		Oid optype = get_opclass_input_type(opclass->values[attoff]);
+		/*
+		 * Load the operator info.  We need this to get the equality operator
+		 * function for the scan key.
+		 */
+		opfamily = get_opclass_family(opclass->values[attoff]);
+		operator= get_opfamily_member(opfamily, optype, optype, BTEqualStrategyNumber);
+		regop = get_opcode(operator);
+
+		if (attoff < indnkeyatts - 1)
+		{
+			/* Initialize segmentby scankeys. */
+			if (key_flags & SEGMENTBY_KEYS)
+			{
+				/* get index attribute name */
+				Form_pg_attribute attr = TupleDescAttr(idxrel->rd_att, attoff);
```

Comment:
Can we end up with index scan with no keys here? Probably better to keep the seq scan for this case, because a full index scan w/o keys is just needlessly slower.

Reply:
I don't understand what you mean here ?
By default there is always an index with segmentby columns on compressed chunk.
Index Oid is fetched from get_compressed_chunk_index() which will always check for presence of segmentby columns in the index, else this function will return InvalidOid.

Reply:
I mean, can `key_flags & SEGMENTBY_KEYS` be false? That would mean we're using an index scan, but without any index conditions.

Reply:
Fixed.